include/picongpu/plugins/xrayScattering/XrayScattering.hpp

```diff
@@ -0,0 +1,846 @@
+                 *   The value retrieved by getDomainSize in
+                 *   CoordinateTransform.hpp is still set to (0,0,0) when the
+                 *   XrayScattering object is initialized.
+                 */
+                probingBeam = pmacc::memory::makeUnique<
+                    beam::XrayScatteringBeam >( );
+                // Set the steps at which the xrayScattering amplitude is
+                // calculated.
+                Environment<>::get( ).PluginConnector( ).setNotificationPeriod(
+                    this,
+                    notifyPeriod
+                );
+                // Set the memory layout in use.
+                std::map< std::string, OutputMemoryLayout > layoutMap;
+                layoutMap["mirror"] = OutputMemoryLayout::MirrorOnNodes;
+                layoutMap["split"] = OutputMemoryLayout::SplitOverNodes;
+                outputLayout = layoutMap.at( memoryLayout );
+
+                mpiRank = gc.getGlobalRank( );
+                isMaster = ( mpiRank == 0 );
+
+                // Prepare amplitude buffer:
+                uint32_t  bufferSize;
+                auto totalNumVectors = numVectors.productOfComponents( );
+                if ( outputLayout == OutputMemoryLayout::MirrorOnNodes )
+                {
+                    // All vectors are stored on every node.
+                    bufferSize =  totalNumVectors ;
+                    // Initiate the additional amplitude storage for the reduce
+                    // operation and initiate it with zeros.
+                    amplitudeMaster.assign( totalNumVectors, complex_X( 0.0 ) );
+                }
+                else
+                {
+                    countRanks = gc.getGpuNodes( ).productOfComponents( );
+                    // Number of scattering vectors in all but last chunk.
+                    // (ceil integer division)
+                    bufferSize = totalNumVectors / countRanks +
+                        ( ( totalNumVectors % countRanks ) != 0 );
+                    // Number of scattering vectors on the last chunk.
+                    resOfVectors = bufferSize - ( bufferSize * countRanks -
+                        totalNumVectors);
+                    // Initiate the additional amplitude storage for receiving
+                    // data and initiate it with zeros.
+                    amplitudeReceive.assign(
+                        bufferSize,
+                        complex_X( 0.0 )
+                    );
+                }
+                // Allocate amplitude buffer.
+                amplitude = pmacc::memory::makeUnique< ComplexBuffer >(
+                    DataSpace< DIM1 >( bufferSize ) );
+                // Initialize, on device, its fields with zero.
+                amplitude->getDeviceBuffer( ).setValue( 0.0 );
+
+                // Go to PIC unit system.
+                q_min = q_min * 1.0e10 * UNIT_LENGTH;
+                q_max = q_max * 1.0e10 * UNIT_LENGTH;
```

Comment:
Perhaps a helper `constexpr` variable for 10^10 with a telling name would help explaining the conversion?